delfin/alert_manager/snmp_validator.py

```diff
@@ -0,0 +1,183 @@
+                error_indication, __, __, __ = cmd_gen.getCmd(
+                    cmdgen.UsmUserData(alert_source['username'],
+                                       authKey=auth_key,
+                                       privKey=privacy_key,
+                                       authProtocol=auth_protocol,
+                                       privProtocol=privacy_protocol,
+                                       securityEngineId=engine_id),
+                    cmdgen.UdpTransportTarget((alert_source['host'],
+                                               alert_source['port']),
+                                              timeout=alert_source[
+                                                  'expiration'],
+                                              retries=alert_source[
+                                                  'retry_num']),
+                    constants.SNMP_QUERY_OID,
+                )
+
+                if 'securityEngineId' in observer_context:
+                    engine_id = observer_context.get('securityEngineId')
+                    alert_source['engine_id'] = binascii.hexlify(
+                        engine_id.asOctets()).decode()
+            else:
+                community_string = cryptor.decode(
+                    alert_source['community_string'])
+                error_indication, __, __, __ = cmd_gen.getCmd(
+                    cmdgen.CommunityData(
+                        community_string,
+                        contextName=alert_source['context_name']),
+                    cmdgen.UdpTransportTarget((alert_source['host'],
+                                               alert_source['port']),
+                                              timeout=alert_source[
+                                                  'expiration'],
+                                              retries=alert_source[
+                                                  'retry_num']),
+                    constants.SNMP_QUERY_OID,
+                )
+
+            if not error_indication:
+                return alert_source
+
+            # Prepare exception with error_indication
+            msg = six.text_type(error_indication)
+        except Exception as e:
+            msg = six.text_type(e)
+
+        # Since validation occur error, raise exception
+        LOG.error("Configuration validation failed with alert source for "
+                  "reason: %s." % msg)
+        raise exception.SNMPConnectionFailed(msg)
+
+    def _handle_validation_error(self, ctxt, storage_id):
+        try:
+            storage = db.storage_get(ctxt, storage_id)
+            alert = {
+                'storage_id': storage['id'],
+                'storage_name': storage['name'],
+                'vendor': storage['vendor'],
+                'model': storage['model'],
+                'serial_number': storage['serial_number'],
+                'alert_id': '19660818',
```

Comment:
Let us use a constant or enum to hold all internal delfin alert IDs, as of now I think we have only this alert.

Reply:
Added a constant for now